src/lib/migration-chunks/validation/tag.ts

```diff
@@ -0,0 +1,114 @@
+import errors from './errors'
+import Tag from '../../entities/tag'
+import Intent from '../../interfaces/intent'
+import { InvalidActionError } from '../../interfaces/errors'
+const tagErrors = errors.tag
+
+interface ValidationContext {
+  remote: Set<string>
+  created: Set<string>
+  deleted: Set<string>
+  toBeCreated: Set<string>
+  remoteTags: Tag[]
+}
+
+interface TagValidation {
+  validate (intent: Intent, context: ValidationContext): string | string[]
+}
+
+class DuplicateCreate implements TagValidation {
+  validate (intent: Intent, context: ValidationContext) {
+    if (!intent.isTagCreate()) {
+      return
+    }
+
+    const tagId = intent.getTagId()
+
+    if (!context.created.has(tagId)) {
+      return
+    }
+
+    return tagErrors.create.TAG_ALREADY_CREATED(tagId)
+  }
+}
+
+class AlreadyExistingCreates implements TagValidation {
+  message = tagErrors.create.TAG_ALREADY_EXISTS
+  validate (intent: Intent, context: ValidationContext) {
+    if (!intent.isTagCreate()) {
+      return
+    }
+
+    const tagId = intent.getTagId()
+
+    if (!context.remote.has(tagId)) {
+      return
+    }
+
+    return tagErrors.create.TAG_ALREADY_EXISTS(tagId)
+  }
+}
+
+const checks: TagValidation[] = [
+  new DuplicateCreate(),
+  new AlreadyExistingCreates()
+]
+
+export default function (intents: Intent[], tags: Tag[]): InvalidActionError[] {
+  const remote = tags.map((tag) => tag.id)
+  const toBeCreated = intents.filter((intent) => intent.isTagCreate()).map((intent) => intent.getTagId())
+
+  let context = {
+    remote: new Set(remote), // all currently (in the current iteration step) existing tags
+    created: new Set<string>(), // all by now (in previous iteration steps) created tags
+    deleted: new Set<string>(), // all by now (in previous iteration steps) deleted tags
+    toBeCreated: new Set(toBeCreated), // all future (in remaining iteration steps) created tags
+    remoteTags: tags
+  }
+
+  let errors = []
+
+  for (const intent of intents) {
+    let error
+
+    for (const check of checks) {
+      error = check.validate(intent, context)
+
+      if (error && error.length) {
+        // proceed with next intent
+        break
+      }
+    }
+
+    if (error) {
+      const errorList = Array.isArray(error) ? error : [error]
+      const invalidActions = errorList.map((error) => ({
+        type: 'InvalidAction',
+        message: error,
+        details: { intent }
+      }))
+
+      errors = errors.concat(invalidActions)
+
+      // do not update context
+      continue
+    }
+
+    const tagId = intent.getTagId()
+
+    if (intent.isTagCreate()) {
+      context.created.add(tagId)
+      context.toBeCreated.delete(tagId)
+      context.deleted.delete(tagId)
+    }
+
+    // TODO
+    // if (intent.isTagDelete()) {
+    //   context.deleted.add(tagId)
+    //   context.remote.delete(tagId)
+    //   context.created.delete(tagId)
+    // }
```

Comment:
What's the issue here?

Reply:
This is leftover code I want to implement in the next pr, I forgot to remove it for this one. Sorry about that!